File: mbopml.py
import requests
import argparse
import xml.etree.ElementTree as ET

def download_following_list(api_key, username, format):
    url = f"https://micro.blog/users/following/{username}"
    headers = {"Authorization": f"Bearer {api_key}"}
    response = requests.get(url, headers=headers)
    
    if response.status_code == 200:
        following_list = response.json()
        opml = generate_opml(following_list, format)
        print(f"{opml}")
    else:
        print(f"Failed to download following list. Status code: {response.status_code}")

def generate_opml(following_list, format):
    root = ET.Element("opml")
    head = ET.SubElement(root, "head")
    body = ET.SubElement(root, "body")
    
    for user in following_list:
        outline = ET.SubElement(body, "outline")
        outline.set("text", user["name"])  # Change outline text to be the feed's title
        outline.set("type", "rss")
        outline.set("title", user["name"])  # Add a new field called "title" and set its value to the feed's title
        if format == "xml":
            outline.set("xmlUrl", f"https://{user['username']}.micro.blog/feed.xml")
        elif format == "json":
            outline.set("xmlUrl", f"https://{user['username']}.micro.blog/feed.json")
    
    tree = ET.ElementTree(root)
    opml_str = ET.tostring(root, encoding="utf-8", method="xml").decode()
    return opml_str

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--api-key", help="Micro.blog API key")
    parser.add_argument("--username", help="Username of the user to download following list")
    parser.add_argument("--format", help="Format of the feed (xml or json)")
    args = parser.parse_args()

    if args.api_key and args.username and args.format:
        download_following_list(args.api_key, args.username, args.format)
    else:
        print("Please provide the Micro.blog API key, username, and format using the --api-key, --username, and --format arguments.")

File: src/mbopml.rs
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION};
use clap::Parser;
use serde_json::Value;
use xml::writer::{EmitterConfig, XmlEvent};

#[derive(Parser)]
#[command(name = "mbopml")]
#[command(about = "A tool to interact with Micro.blog API", long_about = None)]
struct Opt {
    /// Micro.blog API key
    #[arg(long = "api-key")]
    api_key: String,

    /// Username
    #[arg(long = "username")]
    username: String,

    /// Format
    #[arg(long = "format")]
    format: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let opt = Opt::parse();

    let api_key = &opt.api_key;
    let username = &opt.username;
    let format = &opt.format;

    let url = format!("https://micro.blog/users/following/{}", username);
    let mut headers = HeaderMap::new();
    headers.insert(
        AUTHORIZATION,
        HeaderValue::from_str(&format!("Bearer {}", api_key))?,
    );

    let client = reqwest::Client::new();
    let response = client.get(&url).headers(headers).send().await?;

    if response.status().is_success() {
        let following_list: Value = response.json().await?;
        let opml = generate_opml(following_list, format);
        println!("{}", opml);
    } else {
        println!(
            "Failed to download following list. Status code: {}",
            response.status()
        );
    }

    Ok(())
}

fn generate_opml(following_list: Value, format: &str) -> String {
    let mut buffer = Vec::new();
    let mut writer = EmitterConfig::new()
        .perform_indent(true)
        .create_writer(&mut buffer);

    writer.write(XmlEvent::start_element("opml")).unwrap();
    writer.write(XmlEvent::start_element("head")).unwrap();
    writer.write(XmlEvent::end_element()).unwrap(); // end head
    writer.write(XmlEvent::start_element("body")).unwrap();

    for user in following_list.as_array().unwrap() {
        let name = user["name"].as_str().unwrap();
        let username = user["username"].as_str().unwrap();
        let xml_url = match format {
            "xml" => format!("https://{}.micro.blog/feed.xml", username),
            "json" => format!("https://{}.micro.blog/feed.json", username),
            _ => String::new(),
        };
        writer.write(XmlEvent::start_element("outline")).unwrap();
        writer.write(XmlEvent::characters(name)).unwrap();
        writer.write(XmlEvent::start_element("type")).unwrap();
        writer.write(XmlEvent::characters("rss")).unwrap();
        writer.write(XmlEvent::end_element()).unwrap(); // end type
        writer.write(XmlEvent::start_element("title")).unwrap();
        writer.write(XmlEvent::characters(name)).unwrap();
        writer.write(XmlEvent::end_element()).unwrap(); // end title
        writer.write(XmlEvent::start_element("xmlUrl")).unwrap();
        writer.write(XmlEvent::characters(&xml_url)).unwrap();
        writer.write(XmlEvent::end_element()).unwrap(); // end xmlUrl
        writer.write(XmlEvent::end_element()).unwrap(); // end outline
    }

    writer.write(XmlEvent::end_element()).unwrap(); // end body
    writer.write(XmlEvent::end_element()).unwrap(); // end opml

    String::from_utf8(buffer).unwrap()
}
